### PR TITLE
feat: unified Reviewer architecture with self-review framework

### LINKED ISSUE #28: feat: Architecture v3 - Unified Reviewer, Self-Review Framework, Agent Hooks
## Summary

Leverage new Claude Code features (agent hooks, `context: fork`, skill `agent` field) to simplify our architecture:

1. **Consolidate 11 review agents → 1 parameterized `Reviewer`**
2. **Add self-review framework to Coder** with fix-before-returning behavior
3. **Use agent hooks** for lifecycle control and file persistence
4. **DRY review orchestration** - `/implement` calls `/review`

## New Claude Code Features We're Leveraging

| Feature | How We Use It |
|---------|---------------|
| `context: fork` | Run Coder and Reviewer in isolated sub-agent contexts |
| `agent` field in skills | Skills can specify which agent executes them |
| Agent `hooks.Stop` | Self-review in Coder, file persistence in Reviewer |
| `type: prompt` hooks | Intelligent self-review decisions |

## Design Decisions

| Decision | Choice | Rationale |
|----------|--------|-----------|
| Orchestrator context | Main (visible) | User sees progress, can intervene at gates |
| Review gating | Report only | Show findings, user decides |
| Coder → Review transition | Orchestrator spawns | /implement calls /review command |
| Review consolidation | Parameterized Reviewer | Pattern skills exist, methodology shared |
| Self-review behavior | Fix, not just report | Catch issues before external review |
| Review output format | Markdown (.md) | Human readable |
| Migration strategy | Big bang + coexistence | Keep old agents for rollback |

### Current (11 Review Agents)
```
/implement
  ├─→ Coder
  └─→ SecurityReview    ─┐
      ArchitectureReview ├─→ Summary
      PerformanceReview  │
      ComplexityReview   │
      ConsistencyReview  │
      TypescriptReview   │
      TestsReview        │
      DocumentationReview│
      DependenciesReview │
      RegressionReview   │
      DatabaseReview    ─┘

Problem: 11 agents with duplicated methodology
```

### Proposed (1 Parameterized Reviewer)
```
/implement (main context, visible)
  │
  ▼
┌─────────────────────────────────────────────────────────────┐
│                 Coder (context: fork)                        │
│  hooks:                                                      │
│    Stop:                                                     │
│      - type: prompt                                          │
│        prompt: "Self-review using devflow-self-review.       │
│                 Fix all P0/P1 issues. Return when PASS."     │
│                                                              │
│  Output: implementation + self-review report                 │
└─────────────────────────────────────────────────────────────┘
  │
  ▼ (orchestrator calls /review command)
┌─────────────────────────────────────────────────────────────┐
│                      /review                                 │
│  ┌────────────────────────────────────────────────────────┐ │
│  │ Reviewer × N (parallel, context: fork)                  │ │
│  │ focus: security | architecture | performance | ...      │ │
│  │ Loads: devflow-review-methodology + devflow-{focus}-*   │ │
│  │                                                         │ │
│  │ hooks:                                                  │ │
│  │   Stop:                                                 │ │
│  │     - type: command                                     │ │
│  │       command: "# findings already saved to file"       │ │
│  │                                                         │ │
│  │ Output: .docs/reviews/{branch}/{focus}.md               │ │
│  └────────────────────────────────────────────────────────┘ │
│                            │                                 │
│                            ▼                                 │
│  ┌────────────────────────────────────────────────────────┐ │
│  │ Summary (context: fork)                                 │ │
│  │ Reads: .docs/reviews/{branch}/*.md                      │ │
│  │ Output: .docs/reviews/{branch}/SUMMARY.md               │ │
│  └────────────────────────────────────────────────────────┘ │
└─────────────────────────────────────────────────────────────┘
  │
  ▼
┌─────────────────────────────────────────────────────────────┐
│                      USER GATE                               │
│  Shows: Summary + recommendation                             │
│  Options: Create PR / Request fixes / Abort                  │
└─────────────────────────────────────────────────────────────┘
```

### 1. New Skill: `devflow-self-review`

**Iron Law**: FIX BEFORE RETURNING

**8 Pillars** (prioritized):

| Priority | Pillar | Key Checks |
|----------|--------|------------|
| P0 | Design | Fits architecture, interaction logic sound, not over-solving |
| P0 | Functionality | Works as intended, edge cases handled, no race conditions |
| P0 | Security | Input validation, no injection risks, no secrets exposed |
| P1 | Complexity | Understandable in <5 min, single responsibility, no over-abstraction |
| P1 | Error Handling | Explicit handling, Result types where appropriate, no silent failures |
| P1 | Tests | Cover new functionality, would fail if code breaks, edge cases tested |
| P2 | Naming | Descriptive names, comments explain "why" not "what" |
| P2 | Consistency | Follows style guide, matches surrounding patterns |

**Behavior**:
- Evaluate P0 → fix immediately → evaluate P1 → fix → evaluate P2 → fix if time
- If P0 issue unfixable → STOP and report blocker
- Only return when all P0 and P1 PASS

Based on: [Google Engineering Practices](https://google.github.io/eng-practices/review/reviewer/looking-for.html), [Microsoft Engineering Playbook](https://microsoft.github.io/code-with-engineering-playbook/code-reviews/process-guidance/reviewer-guidance/)

### 2. New Agent: `Reviewer`

```yaml
---
name: Reviewer
description: Universal code review agent with parameterized focus
model: sonnet
skills: devflow-review-methodology, devflow-security-patterns,
        devflow-architecture-patterns, devflow-performance-patterns, ...
tools: Read, Grep, Glob, LSP
---
```

**Parameterization**: Invoked with `focus` parameter that determines which pattern skill to apply:
- `focus: security` → applies `devflow-security-patterns`
- `focus: architecture` → applies `devflow-architecture-patterns`
- etc.

**Output**: Writes findings to `.docs/reviews/{branch}/{focus}.md`

### 3. Updated Agent: `Coder`

Add to frontmatter:
```yaml
skills: devflow-core-patterns, devflow-implementation-patterns,
        devflow-git-safety, devflow-self-review
hooks:
  Stop:
    - hooks:
        - type: prompt
          prompt: |
            Before completing, run self-review using devflow-self-review.
            Evaluate each pillar in order. FIX any issues found.
            Only return when all P0 and P1 pillars PASS.
            Output: fixes made + final status.
```

### 4. Updated Command: `/implement`

Change Phase 8-9 (review phases) to:
```markdown
## Phase 8: Code Review

Invoke the /review command to run comprehensive review.

This ensures review orchestration logic is not duplicated.
```

### 5. Updated Command: `/review`

Replace individual review agent spawns with parameterized Reviewer:
```markdown
## Phase 3: Parallel Reviews

Spawn Reviewer agent for each focus area in parallel:

| Focus | Condition |
|-------|-----------|
| security | Always |
| architecture | Always |
| performance | Always |
| complexity | Always |
| consistency | Always |
| tests | Always |
| regression | Always |
| typescript | If .ts/.tsx files changed |
| database | If migration/schema files changed |
| documentation | If docs changed |
| dependencies | If package.json/lock files changed |

Each invocation:
- Agent: Reviewer
- Parameter: focus={focus}
- Output: .docs/reviews/{branch}/{focus}.md
```

## Self-Review vs External Review

| Aspect | Self-Review (Coder) | External Review (Reviewer) |
|--------|---------------------|---------------------------|
| **Purpose** | Catch obvious issues before submitting | Fresh eyes, specialized expertise |
| **Action** | Fix immediately | Report findings |
| **Scope** | 8 pillars, general quality | Domain-specific deep dives |
| **When** | Before returning from Coder | After Coder completes |
| **Output** | Fixed code + report | Findings in .md files |

## Migration Strategy

1. **Create new components** (Reviewer, devflow-self-review)
2. **Update existing** (Coder hooks, /review, /implement)
3. **Keep old review agents** (deprecate but don't delete)
4. **Test end-to-end**
5. **Remove old agents in next release** (after validation)

### Phase 1: Pattern Skills Extraction
- [ ] Extract patterns from `review-architecture.md` → `devflow-architecture-patterns`
- [ ] Extract patterns from `review-performance.md` → `devflow-performance-patterns`
- [ ] Extract patterns from `review-complexity.md` → `devflow-complexity-patterns`
- [ ] Extract patterns from `review-consistency.md` → `devflow-consistency-patterns`
- [ ] Extract patterns from `review-tests.md` → `devflow-tests-patterns`
- [ ] Extract patterns from `review-database.md` → `devflow-database-patterns`
- [ ] Extract patterns from `review-documentation.md` → `devflow-documentation-patterns`
- [ ] Extract patterns from `review-dependencies.md` → `devflow-dependencies-patterns`
- [ ] Extract patterns from `review-regression.md` → `devflow-regression-patterns`

### Phase 2: Self-Review Framework
- [ ] Create `skills/devflow-self-review/SKILL.md` with 8 pillars
- [ ] Update `agents/coder.md` with Stop hook

### Phase 3: Unified Reviewer
- [ ] Create `agents/reviewer.md` with parameterized focus
- [ ] Update `commands/review.md` to use Reviewer
- [ ] Update `agents/review-summary.md` to read .md files

### Phase 4: Pipeline Integration
- [ ] Update `commands/implement.md` to call /review
- [ ] Test end-to-end flow

### Phase 5: Cleanup
- [ ] Add deprecation notice to old review agents
- [ ] Update CLAUDE.md documentation
- [ ] Update README.md if needed

## Pattern Skills Gap Analysis

**Existing pattern skills:**
- `devflow-security-patterns` - EXISTS

**Missing pattern skills needed for Reviewer:**
- `devflow-architecture-patterns` - MISSING
- `devflow-performance-patterns` - MISSING
- `devflow-complexity-patterns` - MISSING
- `devflow-consistency-patterns` - MISSING
- `devflow-tests-patterns` - MISSING
- `devflow-database-patterns` - MISSING
- `devflow-documentation-patterns` - MISSING
- `devflow-dependencies-patterns` - MISSING
- `devflow-regression-patterns` - MISSING

**Decision:** Extract ALL patterns from existing review agents into skills upfront. Each current `review-{type}.md` agent contains specialized knowledge that becomes `devflow-{type}-patterns` skill.

This adds ~9 new skill files but:
1. Makes patterns reusable (Coder can use them too)
2. Single source of truth for each domain
3. Cleaner separation of concerns

## References

- Claude Code hooks documentation
- [Google Engineering Practices](https://google.github.io/eng-practices/review/reviewer/looking-for.html)
- [Microsoft Engineering Playbook](https://microsoft.github.io/code-with-engineering-playbook/code-reviews/process-guidance/reviewer-guidance/)
- Issue #27 (Shepherd hooks design - related but different scope)

### PR DESCRIPTION
**Pull request by dean0x**
## Summary

This PR implements Architecture v3 from Issue #28, fundamentally restructuring how code reviews work in DevFlow:

- **11 → 1**: Consolidate 11 individual review agents into a single parameterized `Reviewer` agent
- **Self-review**: Add 9-pillar self-review framework to `Coder` agent via Stop hook
- **Pattern skills**: Create 10 new pattern skills as shared knowledge libraries
- **DRY orchestration**: `/implement` now calls `/review` instead of duplicating logic

### New Pattern Skills (10 files)

| Skill | Focus Area | Iron Law |
|-------|------------|----------|
| `devflow-architecture-patterns` | SOLID, coupling, layering | VIOLATIONS IN YOUR CHANGES ARE BLOCKING |
| `devflow-performance-patterns` | N+1, algorithms, memory, I/O | MEASURE BEFORE OPTIMIZING |
| `devflow-complexity-patterns` | Cyclomatic complexity, readability | COMPLEXITY IS THE ENEMY OF RELIABILITY |
| `devflow-consistency-patterns` | Pattern violations, simplification | FOLLOW EXISTING PATTERNS |
| `devflow-tests-patterns` | Coverage, quality, brittleness | TESTS MUST VALIDATE BEHAVIOR NOT IMPLEMENTATION |
| `devflow-database-patterns` | Schema, queries, migrations | MIGRATIONS MUST BE REVERSIBLE |
| `devflow-documentation-patterns` | Docs quality, alignment | DOCS MUST MATCH CODE |
| `devflow-dependencies-patterns` | CVEs, versions, licenses | NO VULNERABLE DEPENDENCIES IN PRODUCTION |
| `devflow-regression-patterns` | Lost functionality, broken behavior | PRESERVE EXISTING FUNCTIONALITY |
| `devflow-self-review` | 9-pillar framework | FIX BEFORE RETURNING |

### Unified Reviewer Agent

Single `agents/reviewer.md` with all pattern skills loaded. Focus specified via prompt injection:

```
Task(subagent_type="Reviewer"):
"Review this code focusing exclusively on SECURITY.
Apply patterns from devflow-security-patterns skill.
Follow the 6-step process from devflow-review-methodology."
```

### Coder Self-Review (Stop Hook)

```yaml
hooks:
  Stop:
    - hooks:
        - type: prompt
          prompt: |
            Before completing, run self-review using devflow-self-review skill.
            Evaluate each of the 9 pillars in priority order (P0 -> P1 -> P2).
            FIX any P0 or P1 issues found - do not just report them.
            Only return when all P0 and P1 pillars PASS.
```

**9 Pillars**:
- P0 (MUST pass): Design, Functionality, Security
- P1 (SHOULD pass): Complexity, Error Handling, Tests  
- P2 (Fix if time): Naming, Consistency, Documentation

### Removed

11 individual review agent files:
- `review-architecture.md`, `review-complexity.md`, `review-consistency.md`
- `review-database.md`, `review-dependencies.md`, `review-documentation.md`
- `review-performance.md`, `review-regression.md`, `review-security.md`
- `review-tests.md`, `review-typescript.md`

## Architecture

```
/review (orchestrator)
├─ Spawns 7-11 Reviewer agents in parallel (different focus areas)
├─ Spawns Comment agent (PR inline comments)
├─ Spawns TechDebt agent (backlog tracking)
└─ Spawns Summary agent (merge recommendation)

/implement (orchestrator)
├─ Phase 1-5: Explore + Plan
├─ Phase 6: Coder agent (with self-review via Stop hook)
├─ Phase 7-8: Commit + PR
└─ Phase 9: Calls /review command (DRY)
```

## Test Plan

- [ ] Run `/review` on a branch with mixed file types
- [ ] Verify Reviewer agents spawn with correct focus areas
- [ ] Verify self-review triggers in Coder before completion
- [ ] Verify pattern skills provide correct detection patterns
- [ ] Verify `/implement` calls `/review` instead of duplicating

Closes #28

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: skills/devflow-documentation-patterns/SKILL.md [hunk 1]
```diff
@@ -0,0 +1,352 @@
...
+- [ ] Complex algorithms explained
+- [ ] Comments explain "why", not "what"
+- [ ] README reflects current state
+- [ ] CHANGELOG updated for notable changes
+- [ ] No TODO comments for completed work
+- [ ] Examples work with current API
+
+---
+
+## Integration
```
Comment (dean0x) [inline skills/devflow-documentation-patterns/SKILL.md]: This section seems irrelevant once the skill already loaded what am I missing?

File: skills/devflow-tests-patterns/SKILL.md [hunk 2]
```diff
@@ -0,0 +1,406 @@
...
+|-----------|-------------------|-----------|
+| Business logic | 90%+ | Unit tests |
+| API endpoints | 80%+ | Integration tests |
+| UI components | 70%+ | Component tests |
+| Utilities | 100% | Unit tests |
+| Error paths | 100% | Unit tests |
+
+---
+
+## Integration
```
Comment (dean0x) [inline skills/devflow-tests-patterns/SKILL.md]: This section seems irrelevant once the skill already loaded what am I missing?

File: skills/devflow-self-review/SKILL.md [hunk 3]
```diff
@@ -0,0 +1,488 @@
...
+    └────┬────┘
+         │
+         ▼
+      RETURN
+   WITH REPORT
+```
+
+---
+
+## Integration
```
Comment (dean0x) [inline skills/devflow-self-review/SKILL.md]: This section seems irrelevant once the skill already loaded what am I missing?
